### PR TITLE
*: check memory limit when inserting in-memory pessimistic locks

### PR DESCRIPTION
**Pull request by sticnarf**
### What is changed and how it works?

ref #11452

What's Changed:

This PR adds memory tracing for in-memory pessimistic locks and check memory limit before insertion.

To make the memory tracing more precise, the `map` is made private, so the map can be only changed through public functions which maintains the memory usage.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/txn_ext.rs [hunk 1]
```diff
@@ -45,6 +47,19 @@ impl fmt::Debug for TxnExt {
...
+        "tikv_pessimistic_lock_memory_size",
+        "Total memory size of pessimistic locks in bytes."
+    )
+    .unwrap();
+}
+
+const GLOBAL_MEM_SIZE_LIMIT: usize = 100 << 20; // 100 MiB
+
+// 512 KiB, so pessimistic locks in one region can be proposed in a single command.
+const PEER_MEM_SIZE_LIMIT: usize = 512 << 10;
```
Comment (youjiali1995) [inline components/raftstore/src/store/txn_ext.rs:61]: Are these too small? Is it necessary to make them adjustable?

**Review by youjiali1995 — APPROVED** [review]
BTW, don't forget to adjust `group_locks_by_regions()` after https://github.com/tikv/tikv/pull/11655 is merged.

File: components/raftstore/src/store/txn_ext.rs (hunk 1 above)
Comment (sticnarf) [reply components/raftstore/src/store/txn_ext.rs:61]: Let's first use a small const, because we don't yet split proposing command before leader transfer.

**Comment by sticnarf** [comment]
> BTW, don't forget to adjust `group_locks_by_regions()` after #11655 is merged.

I move this function to `PeerPessimisticLocks` and add test for it. You can take another look.

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**